### ISSUE TITLE
Entering in a br via <shift-enter> in front of specially-crafted content appears to cause several question marks to appear on the next line

### ISSUE THREAD

**Issue by elgatorico** [ckeditor/ckeditor-dev#712]
## Are you reporting a feature request or a bug?
Bug

Entering in a SHIFT-ENTER (for a `<br>`) directly before the "h" in http leads to question marks being appended to the next line (see repo steps).

## Provide detailed reproduction steps (if any)

1. On an instance of the ckeditor, type the following text (add spaces and enters where indicated):

This is a test[space][space]http://www.mhc.ab.ca/Library/Get_Assistance/How-To_Guides.aspx [enter]
Here is some other content.[enter]

2. Set the cursor prior to the h in http and press shift-enter to insert a `<br>`
3. Submit the text to a web method for persisting on a database (at this point, the html already contains the offending question marks)
4. Retrieve the text back from database

### Expected result

The html code entered should be returned:

`<p>This is a test &nbsp;<br> http://www.mhc.ab.ca/Library/Get_Assistance/How-To_Guides.aspx</p>  <p>Here is some other content.</p>  <p>&nbsp;</p>`

### Actual result

The html is returned. but with added question marks:
`<p>This is a test &nbsp;<br> http://www.mhc.ab.ca/Library/Get_Assistance/How-To_Guides.aspx</p>  <p>???????Here is some other content.</p>  <p>&nbsp;</p>`

## Other details

the editor config:
`CKEDITOR.editorConfig = function (config) {	
	config.language = 'en';	
	config.startupFocus = false;
	config.resize_enabled = true;	
    
    config.extraPlugins = 'tableresize,removeRedundantNBSP,quicktable';	
    
    //can resize vertically to a maximum and minimum extent
    config.resize_dir = 'vertical';
    config.resize_maxHeight = 750;
    config.resize_minHeight = 300;

	config.autoGrow_onStartup = true;
	//this is for removing the footer on the editor, copyformatting and so on
	config.removePlugins = 'elementspath,copyformatting,pastefromword';
	//force pasting as plain text
    config.forcePasteAsPlainText = true;	

	config.uiColor = '#FFFFCC';    
		
    config.toolbarGroups = [
        { name: 'clipboard', groups: ['clipboard', 'undo'] },
        { name: 'editing', groups: ['find', 'selection', 'spellchecker', 'editing'] },
        { name: 'forms', groups: ['forms'] },
        { name: 'basicstyles', groups: ['basicstyles', 'cleanup'] },
        { name: 'paragraph', groups: ['list', 'indent', 'blocks', 'align', 'bidi', 'paragraph'] },
        { name: 'links', groups: ['links'] },
        { name: 'insert', groups: ['insert'] },
        { name: 'styles', groups: ['styles'] },
        { name: 'colors', groups: ['colors'] },
        { name: 'tools', groups: ['tools'] },
        { name: 'others', groups: ['others'] },
        { name: 'about', groups: ['about'] },
        { name: 'document', groups: ['mode', 'document', 'doctools'] }
    ];
    config.removeButtons = 'Print,Preview,NewPage,Save,Templates,Paste,PasteFromWord,SelectAll,Scayt,Checkbox,Form,Radio,TextField,Textarea,Select,Button,ImageButton,HiddenField,Strike,CopyFormatting,RemoveFormat,Blockquote,CreateDiv,BidiLtr,BidiRtl,Language,Anchor,Image,Flash,HorizontalRule,Smiley,Iframe,Styles,Format,Font,FontSize,TextColor,BGColor,ShowBlocks,Maximize,About';
};`

* Browser: Test case occurs reliably in Chrome v59.0.3071.115 and v60.0.3112.78. I have not been able to reproduce it in other browsers.
* OS: Windows 10
* CKEditor version: 4.7.1
* Installed CKEditor plugins:
a11yhelp
about
clipboard
colordialog
copyformatting
dialog
div
find
flash
forms
iframe
image
link
liststyle
magicline
pagebreak
pastefromword
preview
quicktable
removeRedundantNBSP
scayt
showblocks
smiley
specialchar
table
tableresize
tableselection
tabletools
tabletoolstoolbar
templates
wsc

**Comment by msamsel** [comment]
Hi,

I'm not able to reproduce this error. I use MacOS with Chrome 59. Can you check this sample: https://codepen.io/msamsel/pen/PKGQbM?editors=1010 I used it to test your code, but problem not occur for me.
Could you check this link and give an info, if the problem exists there also?
Can you also extend 3rd and 4th point:
> * Submit the text to a web method for persisting on a database (at this point, the html already contains the offending question marks)
> * Retrieve the text back from database

What methods from CKEDITOR do you use to obtain data? CKEditor don't have included any database system so 4th point is quite confusing for me.
It would be great, if you could provide some code snippet in some online tool like codepen.io or jsfiddle.net.

**Comment by elgatorico** [comment]
Thanks so much @msamsel 

I looked into this further and it looks like there are some non-ascii characters kinda hidden in the ckeditor.js file. For example, this is what the file looks like in notepad++ encoded in utf-8:

[http://imgur.com/AmZiMhD](url)

And this is what it looks like on notepad++ in the ANSI version:

[http://imgur.com/nuKy3ZH](url)

There are several cases throughout the whole file where this happens. My assumption is that when shift+enter is triggered, non-ascii characters are appended along with `<br>`. While the non-ascii characters are not displayed in an utf-8 scenario, they are visible in an ANSI scenario like in a varchar column in a SQL Server DB. Non-ascii characters are then persisted as ??????? because there is no representation of those characters in the ANSI charset.

Is the minification process throwing these characters in there?

thanks again for your help!

By the way, I did a fresh download of ckeditor when I prepared this case.

**Comment by msamsel** [comment]
Hi,

your problem seems not be with editor, but with writing data into database. I've modified my example: https://codepen.io/msamsel/pen/PKGQbM please check it.
I've made button to get data from editor. As you might see data are get from editor and displayed as text in below `<pre>` element. I cannot make any `????` there.
I suspect that you have some DB settings which wrongly process those data. If problem cannot be reproduce without DB, and only exists when you write data to DB, then unfortunately it seems to be problem of DB not CKEditor and I cannot help you more about it.

**Comment by elgatorico** [comment]
Thanks @msamsel 
I found the issue.
Under certain circumstances, ckeditor appends the Zero-Width Space character (\u200B), this character is not part of the ASCII set and when persisted in an only-English type of column in a DB, the char is stored as ? since there's no interpretation for it in ASCII.

An example of this exist in the file ckeditor\plugins\scayt\plugin.js line 1362:

`/*
					 	CKEDITOR use cke-fillingChar with code "8203" for system processes
					 	If SCAYT have changed DOM content we will use the method "setCustomData"
					 	for providing a link to the new node with special character cke-fillingChar
					 	for this case
					*/
`

or in the ckeditor.js file on line 1064:
`return c?Array(7).join(String.fromCharCode(8203)):String.fromCharCode(8203)`

I think the only option pretty much left for me is to change the column type from VARCHAR TO NVARCHAR to accomodate non-ascii characters. It's not the end of the world, but it would have been nice to know this requirement before hand.

thanks!

**Comment by elgatorico** [comment]
some more info.
This issue was reported in 2015:
https://groups.google.com/forum/#!topic/ckeditor-tickets/a0eCcrRFXG0

and it was closed in 2015 because it was supposed to be fixed on version 4.4.6. I don't necessarily agree that this has been fixed thou:

> Changes (by Reinmar):

 * status:  confirmed => closed
 * resolution:   => fixed


Comment:

 Existence of the ZWS filler inside the editor **is not a bug** (but its
 existence in the data is). This is a fix for a very ugly Blink's and
 Webkit's issues with selections in specific locations - for example inside
 empty inline elements or between inline elements. We create this character
 to make selection stick to it.

 This is and ugly hack, because the bug is ugly. Unfortunately, this means
 that there were and perhaps still are issues. However, we managed to fix
 many of them recently in #12491 and perhaps in #12621 too (both 4.4.6).
 For instance, the issue mentioned in comment:12 is now fixed. The
 originally reported TC is also fixed (see http://jsfiddle.net/J669P/6/).

 To sum up:

 1. The ZWS character is totally valid inside the editor.
 2. The ZWS character should never appear in the data and that issue (the
 original one) was fixed.
 3. There were other issues related to the usage of the ZWS character, but
 all mentioned here where fixed in 4.4.6.

**Comment by msamsel** [comment]
@elgatorico it's very nice research :)
If you have some steps allowing for bug reproduction then feel free to share it with us. Your description with <kbd>Shift</kbd>+<kbd>Enter</kbd> case was not working for me :( I haven't see any ZWS characters in data output.
Generally we use ZWS elements in some cases (mostly to prevent browser of removing some empty elements), but those elements should not appear in data output when you use [`editor.getData()`](https://docs.ckeditor.com/#!/api/CKEDITOR.editor-method-getData) method.
If you find situation when such characters are returned by `getData`, please inform us. This most probably shouldn't happen.

**Comment by HalimG** [comment]
Hi @msamsel,

Not really linked to the original issue but I can reproduce this on Chrome 59 with CK 4.7.0.
On a blank CK, click on "Bold", then "Bulleted list". Then remove "Bulleted list" and remove "Bold".
Your editor is blank and no button is clicked.
editor.getData() will return `"<p>&#8203;&#8203;&#8203;&#8203;&#8203;&#8203;&#8203;</p>"` instead of `""`.

**Comment by msamsel** [comment]
@HalimG does the problem exists in the newest version of CKEditor? I've tried to get this on Chrome 60 and CKEditor 4.7.2 but I get only extra `strong` elements.
Can you prepare some code sample where this issue exists?

**Comment by elgatorico** [comment]
Thank you both. 
I was able to reproduce it with the steps @HalimG provided:
On a blank CK, click on "Bold", then "Bulleted list". Then remove "Bulleted list" and remove "Bold".
Your editor is blank and no button is clicked.

[https://codepen.io/rgamero/pen/vJzKMV](url)

Once you click on "get data", you'll see a bunch of 'u200B', these are ZWSs.

I hope this makes sense, please let me know if you need more clarifications.

**Comment by msamsel** [comment]
@elgatorico can you also provide names and versions of browser and OS, where you were able to reproduce this issue?

**Comment by elgatorico** [comment]
@msamsel I am able to reproduce this on Chrome version 60.0.3112.101 (64bit)

**Comment by elgatorico** [comment]
@msamsel this is a Windows 10 Enterprise box

**Comment by msamsel** [comment]
@elgatorico, @HalimG thank you for the test. I see it now and I can confirm it.
